blog: remote optimization post (#1451)
iterative/dvc.org: content/blog/2020-06-29-optimizing-dvc-remotes.md

---

> title: Optimizing DVC Remotes

Author: probably needs a more interesting title

Reviewer: from the intro think this post is more about "Optimization improvements in DVC 1.0"

> date: 2020-06-29

Author: placeholder date

Reviewer: I think this can be whatever you prefer.

description: |
  An overview of how syncing data to and from remote storage is optimized in DVC.

> picture: 2020-05-04/owl.png

Author: placeholder image

Reviewer: I have the impression if you leave it blank it uses a default img BTW.

Author: I was having issues running the dev server (via `yarn develop`) when `picture` was unset, maybe that's just some problem with my local environment though?

Reviewer: TBH I'm not sure exactly how the blog engine works! You can create a bug report though and Ivan or Roger will probably answer to that 🙂

author: peter_rowlands
---

One of the key features provided by DVC is the ability to efficiently sync
versioned datasets between a user's local machine and
[remote storage](https://dvc.org/doc/command-reference/remote), and version 1.0

> includes several performance optimizations related to syncing data with remotes.

Reviewer: I would start if possible with something like "Our users have presented the need for optimizing remotes blah blah" and give some examples e.g. Discord message screenshots.

Reviewer: Also, minor: I personally prefer "synchronizing" or "syncing". The pronunciation of the latter is questionable, no?

This post provides an overview of some of the ways in which remote performance
has been optimized in DVC.

> In order to sync data between a local machine and remote storage, a tool must
> first do the following (equivalent to
> [status](https://dvc.org/doc/command-reference/status) with `-c/--cloud` in
> DVC), before data can actually be uploaded or downloaded to or from the remote:

Reviewer: ```suggestion
In order to sync data between a local machine and remote storage, we must
first do the following (equivalent to the
`dvc status` command with option `-c` (`--cloud`)
), before data can actually be uploaded or downloaded to or from the remote:
```
> p needs wrapping.

Reviewer: Note that `` `dvc {cmd}` `` is automatically linked to the cmd ref.

> 1. Determine which files are present locally (i.e. in the DVC local cache)

Reviewer: What about external caches? How about just
```suggestion
1. Determine which files are present locally (i.e. in the DVC cache)
```

2. Determine which files are present in the remote

> 3. Determine the difference between the two sets of files

Reviewer: Missing periods `.` in these sentences.

Commonly used cloud sync utilities, such as [rclone](https://rclone.org/), must

> be generalized to support any arbitrary file structure, which can come at the

Reviewer: ```suggestion
be generalized to support any file structure, which can come at the
```

> cost of potential performance, especially when dealing with large datasets
> containing potentially millions of files (or more). In DVC, since we have

Reviewer: ```suggestion
cost of performance, especially when dealing with large datasets.
In DVC, since we have
```
Sentence was already pretty long.

control over the structure of both our local cache and our remote storage, we
leverage this structure to optimize this "status" step in all remote storage

> operations (i.e. `status -c`,
> [push](https://dvc.org/doc/command-reference/push),
> [pull](https://dvc.org/doc/command-reference/pull),
> [fetch](https://dvc.org/doc/command-reference/fetch)). In DVC version 1.0, these

Reviewer: ```suggestion
operations (i.e. `dvc status -c`,
`dvc push`,
`dvc pull`,
`dvc fetch`). In DVC version 1.0, these
```
